classes/notification/PKPNotificationOperationManager.inc.php

```diff
@@ -385,19 +386,87 @@ private function sendNotificationEmail($request, $notification) {
 				$mail->setReplyTo($site->getLocalizedContactEmail(), $site->getLocalizedContactName());
 			}
 
+			$unsubscribeUrl = $this->getUnsubscribeNotificationUrl($request, $notification);
+
 			$mail->assignParams(array(
 				'notificationContents' => $this->getNotificationContents($request, $notification),
 				'url' => $this->getNotificationUrl($request, $notification),
-				'siteTitle' => $context?$context->getLocalizedName():$site->getLocalizedTitle()
+				'siteTitle' => $context?$context->getLocalizedName():$site->getLocalizedTitle(),
+				'notificationUnsubscribeUrl' => $unsubscribeUrl,
 			));
+
 			$mail->addRecipient($user->getEmail(), $user->getFullName());
+
 			if (!$mail->send()) {
 				import('classes.notification.NotificationManager');
 				$notificationMgr = new NotificationManager();
 				$notificationMgr->createTrivialNotification($request->getUser()->getId(), NOTIFICATION_TYPE_ERROR, array('contents' => __('email.compose.error')));
 			}
 		}
 	}
+
+	/**
+	 * Creates and returns a unique string for the given notification, that will be encoded and validated against.
+	 * @param $notification Notification
+	 * @return string
+	 */
+	function createUnsubscribeUniqueKey($notification) {
+		$uniqueKey = 'unsubscribe' . '-' . $notification->getContextId() . '-' . $notification->getUserId() . '-' . $notification->getId();
+
+		return $uniqueKey;
+	}
+
+	/**
+	 * Creates and returns an encoded token that will be used to validate an unsubscribe url.
+	 * @param $notification Notification
+	 * @return string
+	 */
+	function createUnsubscribeToken($notification) {
+		$encodeString = $this->createUnsubscribeUniqueKey($notification);
+
+		$secret = Config::getVar('security', 'signed_page_key_secret', '');
```

Comment:
I wonder if we should reuse the api secret key here rather than adding a new config option. @asmecher do you have an opinion on this?

Reply:
Changed to `api_key_secret`